File: unity/Tests/IntegrationTests.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using NUnit.Framework;
using Newtonsoft.Json.Linq;

namespace Ensoul.Tests
{
    /// <summary>
    /// Integration tests for the Unity C# SDK against a live Docker API stack.
    ///
    /// All tests are skipped when ENSOUL_INTEGRATION_URL is not set.
    ///
    /// Required env vars:
    ///   ENSOUL_INTEGRATION_URL       Base URL, e.g. http://localhost:8000
    ///
    /// Optional env vars:
    ///   ENSOUL_INTEGRATION_USERNAME  Demo username (default: pro-user)
    ///   ENSOUL_INTEGRATION_PASSWORD  Password for the demo user
    ///   ENSOUL_INTEGRATION_DOMAIN    Domain slug for persona CRUD + SSE tests
    /// </summary>
    [TestFixture]
    public class IntegrationTests
    {
        private static readonly string? IntegrationUrl =
            Environment.GetEnvironmentVariable("ENSOUL_INTEGRATION_URL")?.TrimEnd('/');
        private static readonly string IntegrationUsername =
            Environment.GetEnvironmentVariable("ENSOUL_INTEGRATION_USERNAME") ?? "pro-user";
        private static readonly string? IntegrationPassword =
            Environment.GetEnvironmentVariable("ENSOUL_INTEGRATION_PASSWORD");
        private static readonly string? IntegrationDomain =
            Environment.GetEnvironmentVariable("ENSOUL_INTEGRATION_DOMAIN");

        private EnsoulClient _client = null!;
        private EnsoulClient _noAuthClient = null!;
        private string _bearerToken = "";
        private string _testPersonaId = "";
        private bool _personaCreated = false;   // true = we own it (delete in teardown)

        [OneTimeSetUp]
        public async Task OneTimeSetUp()
        {
            if (string.IsNullOrEmpty(IntegrationUrl))
                Assert.Ignore("ENSOUL_INTEGRATION_URL not set");

            // Exchange credentials for a bearer token
            if (!string.IsNullOrEmpty(IntegrationPassword))
            {
                _bearerToken = await ExchangeToken(IntegrationUrl!, IntegrationUsername, IntegrationPassword!);
            }

            var config = new EnsoulConfig(
                baseUrl: IntegrationUrl!,
                apiKey: string.IsNullOrEmpty(_bearerToken) ? "" : null,
                bearerToken: string.IsNullOrEmpty(_bearerToken) ? null : _bearerToken,
                maxRetries: 0
            );
            _client = new EnsoulClient(config);

            var noAuthConfig = new EnsoulConfig(baseUrl: IntegrationUrl!, apiKey: "", maxRetries: 0);
            _noAuthClient = new EnsoulClient(noAuthConfig);

            // Obtain a test persona if domain is configured.
            // Try to create; on ServerException (DB mismatch) fall back to borrowing an existing one.
            if (!string.IsNullOrEmpty(IntegrationDomain) && !string.IsNullOrEmpty(_bearerToken))
            {
                try
                {
                    var persona = await _client.Personas.CreateAsync(
                        name: $"inttest-{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}",
                        domain: IntegrationDomain!
                    );
                    _testPersonaId = persona.Id;
                    _personaCreated = true;
                }
                catch (ServerException)
                {
                    // Persona create failed (e.g. DB schema mismatch) — borrow an existing one
                    try
                    {
                        var page = await _client.Personas.ListAsync(perPage: 1);
                        if (page.Items.Count > 0)
                            _testPersonaId = page.Items[0].Id;
                    }
                    catch { }
                    // _personaCreated stays false — we won't delete in teardown
                }
                catch { }
            }
        }

        [OneTimeTearDown]
        public async Task OneTimeTearDown()
        {
            if (_personaCreated && !string.IsNullOrEmpty(_testPersonaId))
            {
                try { await _client.Personas.DeleteAsync(_testPersonaId); } catch { }
            }
            _client?.Dispose();
            _noAuthClient?.Dispose();
        }

        // ---------------------------------------------------------------------------
        // Helpers
        // ---------------------------------------------------------------------------

        private static async Task<string> ExchangeToken(string baseUrl, string username, string password)
        {
            using var http = new HttpClient();
            var body = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["username"] = username,
                ["password"] = password,
            });
            var resp = await http.PostAsync($"{baseUrl}/v1/auth/token", body);
            if (!resp.IsSuccessStatusCode) return "";
            var json = JObject.Parse(await resp.Content.ReadAsStringAsync());
            return json["access_token"]?.ToString() ?? "";
        }

        private void RequirePassword()
        {
            if (string.IsNullOrEmpty(IntegrationPassword))
                Assert.Ignore("ENSOUL_INTEGRATION_PASSWORD not set");
        }

        private void RequireDomain()
        {
            if (string.IsNullOrEmpty(IntegrationDomain))
                Assert.Ignore("ENSOUL_INTEGRATION_DOMAIN not set");
            RequirePassword();
        }

        // ---------------------------------------------------------------------------
        // Health
        // ---------------------------------------------------------------------------

        [Test]
        public async Task TestHealthCheck()
        {
            using var http = new HttpClient();
            var resp = await http.GetAsync($"{IntegrationUrl}/health");
            Assert.That((int)resp.StatusCode, Is.EqualTo(200));
            var json = JObject.Parse(await resp.Content.ReadAsStringAsync());
            var status = json["status"]?.ToString() ?? "";
            Assert.That(status == "ok" || status == "healthy", $"Unexpected health status: {status}");
            Assert.That(json["version"]?.ToString(), Is.Not.Null.And.Not.Empty);
        }

        // ---------------------------------------------------------------------------
        // Auth
        // ---------------------------------------------------------------------------

        [Test]
        public async Task TestTokenExchange()
        {
            RequirePassword();
            var token = await ExchangeToken(IntegrationUrl!, IntegrationUsername, IntegrationPassword!);
            Assert.That(token, Is.Not.Null.And.Not.Empty);
        }

        [Test]
        public async Task TestAuthMe()
        {
            RequirePassword();
            var user = await _client.Auth.MeAsync();
            Assert.That(user.ConsumerId, Is.Not.Null.And.Not.Empty);
            Assert.That(user.Username, Is.EqualTo(IntegrationUsername));
        }

        [Test]
        public void TestNoCredentialsReturns401()
        {
            Assert.ThrowsAsync<AuthenticationException>(async () =>
                await _noAuthClient.Personas.ListAsync());
        }

        // ---------------------------------------------------------------------------
        // Domains
        // ---------------------------------------------------------------------------

        [Test]
        public async Task TestDomainListReturnsArray()
        {
            RequirePassword();
            var page = await _client.Domains.ListAsync();
            Assert.That(page.Items, Is.Not.Null);
        }

        // ---------------------------------------------------------------------------
        // Personas
        // ---------------------------------------------------------------------------

        [Test]
        public void TestPersonaAvailable()
        {
            RequireDomain();
            Assert.That(_testPersonaId, Is.Not.Null.And.Not.Empty, "No persona available — create failed and no existing personas found");
        }

        [Test]
        public async Task TestPersonaGet()
        {
            RequireDomain();
            var persona = await _client.Personas.GetAsync(_testPersonaId);
            Assert.That(persona.Id, Is.EqualTo(_testPersonaId));
        }

        [Test]
        public async Task TestPersonaListShape()
        {
            RequireDomain();
            var page = await _client.Personas.ListAsync(page: 1, perPage: 5);
            Assert.That(page.Items, Is.Not.Null);
            Assert.That(page.PageNumber, Is.EqualTo(1));
            Assert.That(page.PerPage, Is.EqualTo(5));
        }

        [Test]
        public async Task TestPersonaUpdate()
        {
            RequireDomain();
            if (!_personaCreated)
            {
                Assert.Ignore("Skipping update: using borrowed seeded persona (read-only)");
                return;
            }
            var newName = $"inttest-{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}-upd";
            var updated = await _client.Personas.UpdateAsync(
                _testPersonaId,
                new Dictionary<string, object?> { ["name"] = newName }
            );
            Assert.That(updated.Id, Is.EqualTo(_testPersonaId));
            Assert.That(updated.Name, Is.EqualTo(newName));
        }

        [Test]
        public void TestPersonaNotFound()
        {
            RequirePassword();
            Assert.ThrowsAsync<NotFoundException>(async () =>
                await _client.Personas.GetAsync("00000000-0000-4000-a000-999999999999"));
        }

        // ---------------------------------------------------------------------------
        // SSE Streaming
        // ---------------------------------------------------------------------------

        [Test]
        public async Task TestChatStreamSse()
        {
            RequireDomain();
            using var stream = await _client.Chat.StreamAsync(
                personaId: _testPersonaId,
                message: "Say hello in one word."
            );
            var events = new List<ChatStreamEvent>();
            await foreach (var sseEvent in stream.Events())
            {
                events.Add(SseParser.ParseChatEvent(sseEvent));
            }

            Assert.That(events.Count, Is.GreaterThanOrEqualTo(1), "Expected at least one SSE event");
            var finalEvents = events.FindAll(e => e.IsFinal);
            Assert.That(finalEvents.Count, Is.EqualTo(1), "Expected exactly one final event");
            Assert.That(finalEvents[0].TokenUsage, Is.Not.Null);
        }
    }
}

File: swift/Tests/EnsoulTests/IntegrationTests.swift
/// Integration tests for the Swift SDK against a live Docker API stack.
///
/// All tests are skipped when `ENSOUL_INTEGRATION_URL` is not set.
///
/// Required env vars:
///   ENSOUL_INTEGRATION_URL       Base URL, e.g. http://localhost:8000
///
/// Optional env vars:
///   ENSOUL_INTEGRATION_USERNAME  Demo username (default: pro-user)
///   ENSOUL_INTEGRATION_PASSWORD  Password for the demo user
///   ENSOUL_INTEGRATION_DOMAIN    Domain slug for persona CRUD + SSE tests
///
/// Start the stack before running:
///   cd website && docker compose up -d api postgres redis
///   ENSOUL_INTEGRATION_URL=http://localhost:8000 \
///   ENSOUL_INTEGRATION_PASSWORD=demo-dev-only \
///   swift test --filter IntegrationTests
import XCTest
@testable import Ensoul

@available(iOS 15.0, macOS 12.0, *)
final class IntegrationTests: XCTestCase {

    private static let integrationURL = ProcessInfo.processInfo.environment["ENSOUL_INTEGRATION_URL"]?.trimmingCharacters(in: .init(charactersIn: "/"))
    private static let integrationUsername = ProcessInfo.processInfo.environment["ENSOUL_INTEGRATION_USERNAME"] ?? "pro-user"
    private static let integrationPassword = ProcessInfo.processInfo.environment["ENSOUL_INTEGRATION_PASSWORD"]
    private static let integrationDomain = ProcessInfo.processInfo.environment["ENSOUL_INTEGRATION_DOMAIN"]

    var client: EnsoulClient!
    var noAuthClient: EnsoulClient!
    var bearerToken: String = ""
    var testPersonaId: String = ""
    var personaCreated: Bool = false   // true = we own it (delete in tearDown); false = borrowed

    override func setUp() async throws {
        guard let url = Self.integrationURL, !url.isEmpty else {
            throw XCTSkip("ENSOUL_INTEGRATION_URL not set — skipping integration tests")
        }

        // Exchange credentials for a bearer token
        if let password = Self.integrationPassword, !password.isEmpty {
            bearerToken = try await exchangeToken(baseURL: url, username: Self.integrationUsername, password: password)
        }

        client = EnsoulClient(
            apiKey: bearerToken.isEmpty ? "" : nil,
            baseURL: url,
            bearerToken: bearerToken.isEmpty ? nil : bearerToken,
            maxRetries: 0
        )
        noAuthClient = EnsoulClient(apiKey: "", baseURL: url, maxRetries: 0)

        // Obtain a test persona if domain is configured.
        // Try to create; on ServerError (DB mismatch) fall back to borrowing an existing one.
        if let domain = Self.integrationDomain, !domain.isEmpty, !bearerToken.isEmpty {
            do {
                let persona = try await client.personas.create(
                    name: "inttest-\(Int(Date().timeIntervalSince1970))",
                    domain: domain
                )
                testPersonaId = persona.id
                personaCreated = true
            } catch is ServerError {
                // Persona create failed (e.g. DB schema mismatch) — borrow an existing one
                let page = try? await client.personas.list(perPage: 1)
                testPersonaId = page?.items.first?.id ?? ""
                // personaCreated stays false — we won't delete in tearDown
            }
        }
    }

    override func tearDown() async throws {
        if personaCreated && !testPersonaId.isEmpty {
            try? await client.personas.delete(testPersonaId)
        }
    }

    // MARK: - Helpers

    private func exchangeToken(baseURL: String, username: String, password: String) async throws -> String {
        var request = URLRequest(url: URL(string: "\(baseURL)/v1/auth/token")!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let body = "username=\(username)&password=\(password)"
        request.httpBody = body.data(using: .utf8)
        let (data, _) = try await URLSession.shared.data(for: request)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        return json?["access_token"] as? String ?? ""
    }

    private func requirePassword(file: StaticString = #file, line: UInt = #line) throws {
        guard let pwd = Self.integrationPassword, !pwd.isEmpty else {
            throw XCTSkip("ENSOUL_INTEGRATION_PASSWORD not set")
        }
    }

    private func requireDomain(file: StaticString = #file, line: UInt = #line) throws {
        guard let domain = Self.integrationDomain, !domain.isEmpty else {
            throw XCTSkip("ENSOUL_INTEGRATION_DOMAIN not set")
        }
        try requirePassword()
    }

    // MARK: - Health

    func test_health_check() async throws {
        guard let url = Self.integrationURL else { return }
        let (data, response) = try await URLSession.shared.data(from: URL(string: "\(url)/health")!)
        let httpResp = response as! HTTPURLResponse
        XCTAssertEqual(httpResp.statusCode, 200)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let status = json?["status"] as? String ?? ""
        XCTAssertTrue(status == "ok" || status == "healthy", "Unexpected health status: \(status)")
        XCTAssertNotNil(json?["version"])
    }

    // MARK: - Auth

    func test_token_exchange() async throws {
        try requirePassword()
        guard let url = Self.integrationURL, let password = Self.integrationPassword else { return }
        let token = try await exchangeToken(baseURL: url, username: Self.integrationUsername, password: password)
        XCTAssertFalse(token.isEmpty, "Expected a non-empty bearer token")
    }

    func test_auth_me() async throws {
        try requirePassword()
        let user = try await client.auth.me()
        XCTAssertFalse(user.consumerId.isEmpty)
        XCTAssertEqual(user.username, Self.integrationUsername)
    }

    func test_no_credentials_returns_401() async throws {
        do {
            _ = try await noAuthClient.personas.list()
            XCTFail("Expected AuthenticationError")
        } catch let err as AuthenticationError {
            XCTAssertEqual(err.statusCode, 401)
        }
    }

    // MARK: - Domains

    func test_domain_list_returns_array() async throws {
        try requirePassword()
        let page = try await client.domains.list()
        XCTAssertNotNil(page.items)
    }

    // MARK: - Personas

    func test_persona_available() async throws {
        try requireDomain()
        XCTAssertFalse(testPersonaId.isEmpty, "No persona available — create failed and no existing personas found")
    }

    func test_persona_get() async throws {
        try requireDomain()
        let persona = try await client.personas.get(testPersonaId)
        XCTAssertEqual(persona.id, testPersonaId)
    }

    func test_persona_list_shape() async throws {
        try requireDomain()
        let page = try await client.personas.list(page: 1, perPage: 5)
        XCTAssertGreaterThanOrEqual(page.items.count, 0)
        XCTAssertEqual(page.page, 1)
        XCTAssertEqual(page.perPage, 5)
    }

    func test_persona_update() async throws {
        try requireDomain()
        guard personaCreated else {
            throw XCTSkip("Skipping update: using borrowed seeded persona (read-only)")
        }
        let newName = "inttest-\(Int(Date().timeIntervalSince1970))-upd"
        let updated = try await client.personas.update(testPersonaId, name: newName)
        XCTAssertEqual(updated.id, testPersonaId)
        XCTAssertEqual(updated.name, newName)
    }

    func test_persona_not_found() async throws {
        try requirePassword()
        do {
            _ = try await client.personas.get("00000000-0000-4000-a000-999999999999")
            XCTFail("Expected NotFoundError")
        } catch let err as NotFoundError {
            XCTAssertEqual(err.statusCode, 404)
        }
    }

    // MARK: - SSE Streaming

    func test_chat_stream_sse() async throws {
        try requireDomain()
        let stream = client.chat.stream(
            personaId: testPersonaId,
            message: "Say hello in one word."
        )
        var events: [ChatStreamEvent] = []
        for try await rawEvent in stream {
            if let event = try? parseChatEvent(rawEvent) {
                events.append(event)
            }
        }
        XCTAssertGreaterThanOrEqual(events.count, 1, "Expected at least one SSE event")
        let finalEvents = events.filter { $0.isFinal }
        XCTAssertEqual(finalEvents.count, 1, "Expected exactly one final event")
        XCTAssertNotNil(finalEvents.first?.tokenUsage)
    }
}
